File: wisefy/src/androidTest/java/com/isupatches/wisefy/GetSavedNetworksTests.java
package com.isupatches.wisefy;


import android.net.wifi.WifiConfiguration;
import com.isupatches.wisefy.callbacks.GetSavedNetworksCallbacks;
import com.isupatches.wisefy.constants.WiseFyCodes;
import org.junit.Test;
import java.util.ArrayList;
import java.util.List;
import static com.isupatches.wisefy.base.TestUtils.TEST_SSID;
import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.fail;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.after;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.timeout;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;


public class GetSavedNetworksTests extends BaseAndroidJUnit4TestClass {

    @Test
    public void sync_failure_missingPrerequisite() {
        missingPrerequisite();
        assertEquals(null, mWiseFy.getSavedNetworks());
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).getConfiguredNetworks();
    }

    @Test
    public void sync_success() {
        List<WifiConfiguration> wifiList = new ArrayList<>();
        WifiConfiguration mWiFiConfiguration = new WifiConfiguration();
        mWiFiConfiguration.SSID = TEST_SSID;
        wifiList.add(mWiFiConfiguration);
        when(mMockWiFiManager.getConfiguredNetworks()).thenReturn(wifiList);

        assertEquals(wifiList, mWiseFy.getSavedNetworks());
        verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).getConfiguredNetworks();
    }

    @Test
    public void async_failure_missingPrerequisite() {
        missingPrerequisite();
        GetSavedNetworksCallbacks mockCallbacks = mock(GetSavedNetworksCallbacks.class);
        mWiseFy.getSavedNetworks(mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).getSavedNetworksWiseFyFailure(WiseFyCodes.MISSING_PREREQUISITE);
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).getConfiguredNetworks();
    }

    @Test
    public void async_failure_missingPrerequisite_nullCallback() {
        missingPrerequisite();
        try {
            mWiseFy.getSavedNetworks((GetSavedNetworksCallbacks) null);
            verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).getConfiguredNetworks();
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void async_failure_nullSavedNetworks() {
        when(mMockWiFiManager.getConfiguredNetworks()).thenReturn(null);
        GetSavedNetworksCallbacks mockCallbacks = mock(GetSavedNetworksCallbacks.class);
        mWiseFy.getSavedNetworks(mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).noSavedNetworksFound();
    }

    @Test
    public void async_failure_nullSavedNetworks_nullCallback() {
        when(mMockWiFiManager.getConfiguredNetworks()).thenReturn(null);
        try {
            mWiseFy.getSavedNetworks((GetSavedNetworksCallbacks) null);
            verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).getConfiguredNetworks();
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void async_failure_emptyConfiguredNetworks() {
        when(mMockWiFiManager.getConfiguredNetworks()).thenReturn(new ArrayList<WifiConfiguration>());
        GetSavedNetworksCallbacks mockCallbacks = mock(GetSavedNetworksCallbacks.class);
        mWiseFy.getSavedNetworks(mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).noSavedNetworksFound();
    }

    @Test
    public void async_failure_emptyConfiguredNetworks_nullCallback() {
        when(mMockWiFiManager.getConfiguredNetworks()).thenReturn(new ArrayList<WifiConfiguration>());
        try {
            mWiseFy.getSavedNetworks((GetSavedNetworksCallbacks) null);
            verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).getConfiguredNetworks();
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void async_success() {
        List<WifiConfiguration> wifiList = new ArrayList<>();
        WifiConfiguration mWiFiConfiguration = new WifiConfiguration();
        mWiFiConfiguration.SSID = TEST_SSID;
        wifiList.add(mWiFiConfiguration);
        when(mMockWiFiManager.getConfiguredNetworks()).thenReturn(wifiList);

        GetSavedNetworksCallbacks mockCallbacks = mock(GetSavedNetworksCallbacks.class);
        mWiseFy.getSavedNetworks(mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).retrievedSavedNetworks(wifiList);
        verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).getConfiguredNetworks();
    }

    @Test
    public void async_success_nullCallback() {
        List<WifiConfiguration> wifiList = new ArrayList<>();
        WifiConfiguration mWiFiConfiguration = new WifiConfiguration();
        mWiFiConfiguration.SSID = TEST_SSID;
        wifiList.add(mWiFiConfiguration);
        when(mMockWiFiManager.getConfiguredNetworks()).thenReturn(wifiList);

        try {
            mWiseFy.getSavedNetworks((GetSavedNetworksCallbacks) null);
            verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).getConfiguredNetworks();
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void sync_failure_withRegex_missingPrerequisite() {
        missingPrerequisite();
        assertEquals(null, mWiseFy.getSavedNetworks(TEST_SSID));
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).getConfiguredNetworks();
    }

    @Test
    public void sync_success_withRegex() {
        List<WifiConfiguration> wifiList = new ArrayList<>();
        WifiConfiguration mWiFiConfiguration = new WifiConfiguration();
        mWiFiConfiguration.SSID = TEST_SSID;
        wifiList.add(mWiFiConfiguration);
        when(mMockWiFiManager.getConfiguredNetworks()).thenReturn(wifiList);

        assertEquals(wifiList, mWiseFy.getSavedNetworks(TEST_SSID));
        verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).getConfiguredNetworks();
    }

    @Test
    public void async_failure_withRegex_missingPrerequisite() {
        missingPrerequisite();

        GetSavedNetworksCallbacks mockCallbacks = mock(GetSavedNetworksCallbacks.class);
        mWiseFy.getSavedNetworks(TEST_SSID, mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).getSavedNetworksWiseFyFailure(WiseFyCodes.MISSING_PREREQUISITE);
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).getConfiguredNetworks();
    }

    @Test
    public void async_failure_withRegex_missingPrerequisite_nullCallback() {
        missingPrerequisite();

        try{
            mWiseFy.getSavedNetworks(TEST_SSID, null);
            verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).getConfiguredNetworks();
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void async_failure_withRegex_nullSavedNetworks() {
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworksMatchingRegex(anyString())).thenReturn(null);
        GetSavedNetworksCallbacks mockCallbacks = mock(GetSavedNetworksCallbacks.class);
        mWiseFy.getSavedNetworks(TEST_SSID, mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).noSavedNetworksFound();
    }

    @Test
    public void async_failure_withRegex_nullSavedNetworks_nullCallback() {
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworksMatchingRegex(anyString())).thenReturn(null);
        try {
            mWiseFy.getSavedNetworks(TEST_SSID, null);
            verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).getConfiguredNetworks();
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void async_failure_withRegex_emptyConfiguredNetworks() {
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworksMatchingRegex(anyString())).thenReturn(new ArrayList<WifiConfiguration>());
        GetSavedNetworksCallbacks mockCallbacks = mock(GetSavedNetworksCallbacks.class);
        mWiseFy.getSavedNetworks(TEST_SSID, mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).noSavedNetworksFound();
    }

    @Test
    public void async_failure_withRegex_emptyConfiguredNetworks_nullCallback() {
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworksMatchingRegex(anyString())).thenReturn(new ArrayList<WifiConfiguration>());
        try {
            mWiseFy.getSavedNetworks(TEST_SSID, null);
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void async_success_withRegex() {
        List<WifiConfiguration> wifiList = new ArrayList<>();
        WifiConfiguration mWiFiConfiguration = new WifiConfiguration();
        mWiFiConfiguration.SSID = TEST_SSID;
        wifiList.add(mWiFiConfiguration);

        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworksMatchingRegex(anyString())).thenReturn(wifiList);

        GetSavedNetworksCallbacks mockCallbacks = mock(GetSavedNetworksCallbacks.class);
        mWiseFy.getSavedNetworks(TEST_SSID, mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).retrievedSavedNetworks(wifiList);
    }

    @Test
    public void async_success_withRegex_nullCallbacks() {
        List<WifiConfiguration> wifiList = new ArrayList<>();
        WifiConfiguration mWiFiConfiguration = new WifiConfiguration();
        mWiFiConfiguration.SSID = TEST_SSID;
        wifiList.add(mWiFiConfiguration);

        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworksMatchingRegex(anyString())).thenReturn(wifiList);

        try {
            mWiseFy.getSavedNetworks(TEST_SSID, null);
            verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).getConfiguredNetworks();
        } catch (NullPointerException npe) {
            fail();
        }
    }
}

File: wisefy/src/androidTest/java/com/isupatches/wisefy/main/GetSavedNetworksTests.kt
package com.isupatches.wisefy.main

import com.isupatches.wisefy.BaseAndroidJUnit4TestClass
import com.isupatches.wisefy.TEST_SSID
import com.isupatches.wisefy.VERIFICATION_SUCCESS_TIMEOUT
import com.isupatches.wisefy.callbacks.GetSavedNetworksCallbacks
import com.isupatches.wisefy.constants.MISSING_PARAMETER
import org.junit.Assert.assertEquals
import org.junit.Test
import org.mockito.Mockito.mock
import org.mockito.Mockito.timeout
import org.mockito.Mockito.verify

/**
 * Tests the ability to retrieve a list of saved networks on a device.
 *
 * @author Patches
 */
internal class GetSavedNetworksTests : BaseAndroidJUnit4TestClass() {

    @Test fun sync_failure_prechecks() {
        mockWiseFyPrechecksUtil.getSavedNetworks_failure()
        assertEquals(null, wiseFy.getSavedNetworks())
        verificationUtil.didNotTryToGetSavedNetworks()
    }

    @Test fun sync_success() {
        mockWiseFyPrechecksUtil.getSavedNetworks_success()
        val savedNetworks = mockNetworkUtil.savedNetworks()
        assertEquals(savedNetworks, wiseFy.getSavedNetworks())
        verificationUtil.triedToGetSavedNetworks()
    }

    @Test fun async_failure_prechecks() {
        mockWiseFyPrechecksUtil.getSavedNetworks_failure()
        val mockCallbacks = mock(GetSavedNetworksCallbacks::class.java)
        wiseFy.getSavedNetworks(mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).wisefyFailure(MISSING_PARAMETER)
        verificationUtil.didNotTryToGetSavedNetworks()
    }

    @Test fun async_failure_prechecks_nullCallback() {
        mockWiseFyPrechecksUtil.getSavedNetworks_failure()
        nullCallbackUtil.callGetSavedNetworks()
        verificationUtil.didNotTryToGetSavedNetworks()
    }

    @Test fun async_failure_nullSavedNetworks() {
        mockWiseFyPrechecksUtil.getSavedNetworks_success()
        mockNetworkUtil.getConfiguredNetworks_null()
        val mockCallbacks = mock(GetSavedNetworksCallbacks::class.java)
        wiseFy.getSavedNetworks(mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).noSavedNetworksFound()
        verificationUtil.triedToGetSavedNetworks()
    }

    @Test fun async_failure_nullSavedNetworks_nullCallback() {
        mockWiseFyPrechecksUtil.getSavedNetworks_success()
        mockNetworkUtil.getConfiguredNetworks_null()
        nullCallbackUtil.callGetSavedNetworks()
        verificationUtil.triedToGetSavedNetworks()
    }

    @Test fun async_failure_emptyConfiguredNetworks() {
        mockWiseFyPrechecksUtil.getSavedNetworks_success()
        mockNetworkUtil.getConfiguredNetworks_emptyList()
        val mockCallbacks = mock(GetSavedNetworksCallbacks::class.java)
        wiseFy.getSavedNetworks(mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).noSavedNetworksFound()
        verificationUtil.triedToGetSavedNetworks()
    }

    @Test fun async_failure_emptyConfiguredNetworks_nullCallback() {
        mockWiseFyPrechecksUtil.getSavedNetworks_success()
        mockNetworkUtil.getConfiguredNetworks_emptyList()
        nullCallbackUtil.callGetSavedNetworks()
        verificationUtil.triedToGetSavedNetworks()
    }

    @Test fun async_success() {
        mockWiseFyPrechecksUtil.getSavedNetworks_success()
        val savedNetworks = mockNetworkUtil.savedNetworks()
        val mockCallbacks = mock(GetSavedNetworksCallbacks::class.java)
        wiseFy.getSavedNetworks(mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).retrievedSavedNetworks(savedNetworks)
        verificationUtil.triedToGetSavedNetworks()
    }

    @Test fun async_success_nullCallback() {
        mockWiseFyPrechecksUtil.getSavedNetworks_success()
        mockNetworkUtil.savedNetworks()
        nullCallbackUtil.callGetSavedNetworks()
        verificationUtil.triedToGetSavedNetworks()
    }

    @Test fun sync_failure_prechecks_withRegex() {
        mockWiseFyPrechecksUtil.getSavedNetworks_failure()
        assertEquals(null, wiseFy.getSavedNetworks(TEST_SSID))
        verificationUtil.didNotTryToGetSavedNetworks()
    }

    @Test fun sync_success_withRegex() {
        mockWiseFyPrechecksUtil.getSavedNetworks_success()
        val savedNetwork = mockWiseFySearchUtil.findSavedNetworksMatchingRegex_success()
        assertEquals(savedNetwork, wiseFy.getSavedNetworks(TEST_SSID))
    }

    @Test fun async_failure_prechecks_withRegex() {
        mockWiseFyPrechecksUtil.getSavedNetworks_failure()
        val mockCallbacks = mock(GetSavedNetworksCallbacks::class.java)
        wiseFy.getSavedNetworks(TEST_SSID, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).wisefyFailure(MISSING_PARAMETER)
    }

    @Test fun async_failure_prechecks_withRegex_nullCallback() {
        mockWiseFyPrechecksUtil.getSavedNetworks_failure()
        nullCallbackUtil.callGetSavedNetworks_withRegex(TEST_SSID)
        wiseFy.getSavedNetworks(TEST_SSID, null)
    }

    @Test fun async_failure_withRegex_nullSavedNetworks() {
        mockWiseFyPrechecksUtil.getSavedNetworks_success()
        mockWiseFySearchUtil.findSavedNetworksByRegex_null()
        val mockCallbacks = mock(GetSavedNetworksCallbacks::class.java)
        wiseFy.getSavedNetworks(TEST_SSID, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).noSavedNetworksFound()
    }

    @Test fun async_failure_withRegex_nullSavedNetworks_nullCallback() {
        mockWiseFyPrechecksUtil.getSavedNetworks_success()
        mockWiseFySearchUtil.findSavedNetworksByRegex_null()
        nullCallbackUtil.callGetSavedNetworks_withRegex(TEST_SSID)
    }

    @Test fun async_failure_withRegex_emptyConfiguredNetworks() {
        mockWiseFyPrechecksUtil.getSavedNetworks_success()
        mockWiseFySearchUtil.findSavedNetworksByRegex_emptyList()
        val mockCallbacks = mock(GetSavedNetworksCallbacks::class.java)
        wiseFy.getSavedNetworks(TEST_SSID, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).noSavedNetworksFound()
    }

    @Test fun async_failure_withRegex_emptyConfiguredNetworks_nullCallback() {
        mockWiseFyPrechecksUtil.getSavedNetworks_success()
        mockWiseFySearchUtil.findSavedNetworksByRegex_emptyList()
        nullCallbackUtil.callGetSavedNetworks_withRegex(TEST_SSID)
    }

    @Test fun async_success_withRegex() {
        mockWiseFyPrechecksUtil.getSavedNetworks_success()
        val savedNetworks = mockWiseFySearchUtil.findSavedNetworksMatchingRegex_success()
        val mockCallbacks = mock(GetSavedNetworksCallbacks::class.java)
        wiseFy.getSavedNetworks(TEST_SSID, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).retrievedSavedNetworks(savedNetworks)
    }

    @Test fun async_success_withRegex_nullCallbacks() {
        mockWiseFyPrechecksUtil.getSavedNetworks_success()
        mockWiseFySearchUtil.findSavedNetworksMatchingRegex_success()
        nullCallbackUtil.callGetSavedNetworks_withRegex(TEST_SSID)
    }
}
